### PR TITLE
Implement EIC on thumbv7em

### PR DESCRIPTION
**Pull request by twitchyliquid64**
This PR implements everything you need to fire interrupts when a button is pressed. This has been tested on the wio terminal (samd51).

Mad props to jessebraham@ for implementing the low levels here.

I havent got an implementation for thumb6, as I had to make big changes to get v7 working, but it should be mostly copy-paste if someone has hardware.

### CODE REVIEW

**Comment by twitchyliquid64** [comment]
Hmmm, any idea why its failing for that one chip in particular? Those types should exist ...

File: hal/src/common/thumbv7em/eic/pin.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    }
+
+    $(
+        $(#[$attr])*
+        impl<MODE> EicPin<[<$PadType $num>]<gpio::$PinType<gpio::PfA>>> for gpio::$PinType<MODE> {
+            fn into_ei(self, port: &mut Port) -> [<$PadType $num>]<gpio::$PinType<gpio::PfA>> {
+                [<$PadType $num>]::new(self.into_function(port))
+            }
+        }
+        impl<MODE> ExternalInterrupt for &gpio::$PinType<MODE> {
```
Comment (jessebraham) [inline hal/src/common/thumbv7em/eic/pin.rs:136]:
```suggestion
        $(#[$attr])*
        impl<MODE> ExternalInterrupt for &gpio::$PinType<MODE> {
```

**Comment by jessebraham** [comment]
Credit where credit is due, thanks to @keithduncan for the original EIC implementation which I built on. Eventually I'll get around to adapting this version for `thumbv6m` and PRing it.

**Comment by twitchyliquid64** [comment]
:shipit:

**Review by jessebraham — APPROVED** [review]
LGTM, thanks again!

**Pull request merged**